doc/design/memory.md

```diff
@@ -0,0 +1,1595 @@
+  the space has no associated fromspace, it can be destroyed as soon as its
+  unreachable stack is empty.  In practice this seldom occurs because
+  copy-collection commences well before the space becomes empty.
+- The space is capable of containing `size` bytes of value data.  Space
+  utilization is `allocated / size`; copy-collection commences when utilization
+  drops below a chosen threshold.
+
+### Global value layout
+
+Global and local values have the same layout, except that global value headers
+are prefixed with an additional word for refcounting metadata.  Following are
+the valid states for the global value header word.
+
+Global value header bit encodings:
+
+- **c:** Reference count.
+- **z:** Finalizable if 1.  The finalizer function is a low-level runtime
+  function associated with the type, and resurrection is impossible.  Immutable
+  value finalization is intended primarily for runtime implementation of file
+  descriptor management and other similar use cases, rather than as a Hemlock
+  language facility.
+- **s:** Non-`NULL` per space `unreachable` stack linkage.
+- **f:** Forwarding pointer, overlaid with reference count so that the
+  transition from refcounting to forwarding is atomic.
+
+**Global value header word:**
+
+```
+Non-zero refcount:
+
+  cccccccc cccccccc cccccccc cccccccc cccccccc cccccccc cccccccc ccccc0z1
+
+Forwarding pointer:
+
+  ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff fffff000
+
+Zero refcount, deallocated, in space's unreachable value stack:
+
+  ssssssss ssssssss ssssssss ssssssss ssssssss ssssssss ssssssss sssss0z0
+
+Zero refcount, deallocated, decrefing complete:
+
+  00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
+```
+
+### Space metadata lookup
+
+As described above, various space metadata updates occur during deallocation and
+lazy suspension execution.  This requires a function that returns a pointer to
+the associated space metadata, given a value pointer.  A global radix tree
+maintains adequate state to perform this operation.  In order to limit radix
+tree overhead, spaces conform to special size and alignment constraints.  The
+space allocation quantum is a chosen power of 2, e.g. 2 MiB (2<sup>21</sup>).
+All spaces are sized as a whole multiple of the quantum, and they are aligned to
+a quantum boundary.  This allows the radix tree to ignore the least significant
+lg(quantum) address bits, and each space typically only requires one entry in
+the radix tree.  (The extremely unusual exception is for a multi-quantum space
+that contains a huge value, and interior references to that value allow
+references to span multiple quanta.)
+
+### Generation coalescing
+
+The nursery generation grows until its space cannot serve an allocation request,
+at which point it becomes an aging generation.  Aside from allocation via lazy
+suspension execution, a generation cannot grow after it transitions to aging.
```

Comment:
What do we do if there's enough data in the lazy space to require more than one space after copy/collect?

Reply:
Suppose a space is 2 MiB.  Our page allocator isn't constrained to allocate a single 2 MiB region.  If the lazy space(s) push the resulting combined space past the 2 MiB threshold, we just allocate a multiple of the normal space size.  This same approach applies to huge allocations which don't fit in a normally sized space.